lib/elixir/lib/kernel.ex

```diff
@@ -3631,58 +3631,120 @@ defmodule Kernel do
   end
 
   @doc """
-  Range creation operator. Returns a range with the specified `first` and `last` integers.
+  Creates a range from `first` to `last`.
 
-  If last is larger than first, the range will be increasing from
-  first to last. If first is larger than last, the range will be
-  decreasing from first to last. If first is equal to last, the range
-  will contain one element, which is the number itself.
+  If last is more than first, the range will be increasing from
+  first to last. If first is equal to last, the range will contain
+  one element, which is the number itself.
+
+  If first is less than last, the range will be decreasing from first
+  to last, albeit this behaviour is deprecated. Instead prefer to
+  explicitly list the step with `first..last//-1`.
 
   ## Examples
 
       iex> 0 in 1..3
       false
-
-      iex> 1 in 1..3
-      true
-
       iex> 2 in 1..3
       true
 
-      iex> 3 in 1..3
-      true
+      iex> Enum.to_list(1..3)
+      [1, 2, 3]
 
   """
   defmacro first..last do
     case bootstrapped?(Macro) do
       true ->
         first = Macro.expand(first, __CALLER__)
         last = Macro.expand(last, __CALLER__)
+        validate_range!(first, last)
+        # TODO: Deprecate me inside match in all occasions
+        # TODO: Deprecate me inside guard when sides are not integers
         range(__CALLER__.context, first, last)
 
       false ->
         range(__CALLER__.context, first, last)
     end
   end
 
-  defp range(_context, first, last) when is_integer(first) and is_integer(last) do
-    {:%{}, [], [__struct__: Elixir.Range, first: first, last: last]}
+  defp range(_context, first, last) when is_integer(first) and is_integer(last),
+    # TODO: Deprecate me if step is negative
+    step = if first <= last, do: 1, else: -1
+    {:%{}, [], [__struct__: Elixir.Range, first: first, last: last, step: step]}
+  end
+
+  defp range(nil, first, last) do
+    quote(do: Elixir.Range.new(unquote(first), unquote(last)))
+  end
+
+  defp range(_, first, last) do
+    step = quote do
+       min(unquote(first) <= unquote(last) and 1, unquote(first) > unquote(last) and -1)
+    end
+
+    {:%{}, [], [__struct__: Elixir.Range, first: first, last: last, step: step]}
+  end
+
+  @doc """
+  Creates a range from `first` to `last` with `step`.
+
+  ## Examples
+
+      iex> 0 in 1..3//1
+      false
+      iex> 2 in 1..3//1
+      true
+      iex> 2 in 1..3//2
+      false
+
+      iex> Enum.to_list(1..3//1)
+      [1, 2, 3]
+      iex> Enum.to_list(1..3//2)
+      [1, 2, 3]
+      iex> Enum.to_list(3..1//-1)
+      [3, 2, 1]
```

Comment:
How we express things in code will not always translate to a natural language in all possible natural languages. But I believe this definition to be logically quite good, especially because once we added steps I was able to simplify some of our traversal code.